Load JavaScript From PayPal Asynchronously

tubbo opened the pull request:
<img width="1218" alt="Screen Shot 2020-07-22 at 12 57 23 PM" src="https://user-images.githubusercontent.com/113026/88205525-ed429080-cc1a-11ea-8b5c-7b8894348549.png">

Using the `async` attribute in the `<script>` tag that loads PayPal's JavaScript code, Workarea can now prevent it blocking the page in case of a failure, and improve load time performance to boot.

Fixes #10

Commit:
Load JavaScript From PayPal Asynchronously

Using the `async` attribute in the `<script>` tag that loads PayPal's
JavaScript code, Workarea can now prevent it blocking the page in case
of a failure, and improve load time performance to boot.

PAYPAL-7

Fixes #10

mttdffy commented:
Does this not interfere with the workarea modules that try to use the `window.paypal` object? What if it loads but after those modules try to instantiate? They're written to fail gracefully if `window.paypal` is undefined but thats there in case of total failure, not just a delay in loading. We ideally need a way to know when paypal finishes loading to then trigger the modules that need it.

Commit:
Address Race Condition When Using PayPal's SDK

It's possible that our modules that use the PayPal SDK will be
initialized prior to `window.paypal` being available on the page, now
that the JS is loaded in asynchronously, causing unnecessary loss in
functionality. Additionally, there doesn't seem to be a way to bind to
any kind of event given off by the SDK on `window` that would allow us
to know whether the SDK is truly loaded or not. To address this, when
the `window.paypal` object is not available yet, and there's a
`<script>` tag on the page loading the PayPal SDK (indicating that we
are indeed expecting the SDK to be there), the PayPal modules will
re-try initializing every 5 seconds until `window.paypal` is available.

PAYPAL-7

tubbo commented:
@mttdffy Indeed, after setting `async` on the `<script>` tag, my first load of the page did not display the PayPal button on the cart page. Since the `async` property also caches the request once it's done, a simple page refresh got the button back. But I can see how this would be an issue in the real world.

I might be down the wrong path here, but I can't see anywhere in [PayPal's documentation](https://developer.paypal.com/docs/checkout/troubleshoot/performance/#instant-render) that indicates any events are being broadcast to the `window` or `document` object once the PayPal SDK is done loading. So if that's true, we have a few options:

**Option 1, Poll for changes:** If a PayPal `<script>` tag is in the DOM, we will assume that the PayPal SDK is expected to be loaded at some point, so we'll use `setTimeout` to call `init()` every N seconds until `window.paypal` is available to us. This could also be a config that's loaded from ERB, but I figured just testing for the tag's existence would give us the same result. I actually tried this locally and it not only works, it also doesn't seem to block the rest of the page load even when I clear my cache. So that's nice. 
**Option 3, Load the SDK from a JS module:** This is a pattern that we use on other plugins, such as Listrak, and involves sending a custom event to `window` that the rest of the PayPal modules can listen on, and wait to initialize the module until the SDK is available. By far the cleanest approach, but definitely requires the most amount of changes. For example, I'll have to either refactor or move the code out of `PaypalHelper` that determines the URL and attributes for the `<script>` tag. I haven't tried this yet but I'm pretty sure it will work, as we've done it a few times in the past.

To test this stuff out, I'm providing bogus credentials in config. This simulates `window.paypal` not loading in time, as well as the SDK being unavailable because PayPal is down.

Choose Your Fate, Warrior.

mttdffy commented:
@tubbo pretty curious what invisible option 2 is. 

Uh.. option 1 sounds similar to what I was thinking. Not sure how terrible that would be from a cat jesus perspective. There is a lot of critical info getting put together in that helper method that I would rather not mess with.

bencrouse commented:
I had the same initial concern, but they make a good point about the real world in their docs: "For a bonus performance boost, load the PayPal script asynchronously on a page that precedes the Checkout page. This approach pre-caches the script, making future loads/renders instantaneous"

Options 1 or 3 seem like easy enough insurance against this not being the case, but will need to review option 2.

tubbo commented on app/assets/javascripts/workarea/storefront/paypal/modules/paypal_buttons.js:
```diff
@@ -57,9 +57,17 @@ WORKAREA.registerModule('paypalButtons', (function () {
-            var $buttonContainer = $('#paypal-button-container', $scope);
+            var $buttonContainer = $('#paypal-button-container', $scope),
+                paypalNotLoaded = window.paypal === undefined,
+                paypalExpectedToLoad = !_.isEmpty(
```
According to the `include_paypal_javascript_tag` helper method, the `<script>` tag won't even render if the gateway is not configured.

mttdffy commented on app/assets/javascripts/workarea/storefront/paypal/modules/paypal.js:
```diff
@@ -0,0 +1,54 @@
+                paypalExpectedToLoad = !_.isEmpty($sdk);
+
+            if (!paypalLoaded && paypalExpectedToLoad) {
+                window.setTimeout(init, 5000);
```
is 5 seconds too long? Also, this should probably stop checking after so long. If the script becomes unavailable and never loads then this just keeps running. I think maybe 2 second timeout? With maybe a limit of doing that 5-10 times? If after 20(ish) seconds its not loaded its probably not going to.

bencrouse reviewed (changes requested):
See @mttdffy 's comment

mttdffy commented on config/initializers/append_points.rb:
```diff
@@ -25,6 +25,7 @@
 
 Workarea::Plugin.append_javascripts(
   'storefront.modules',
+  'workarea/storefront/paypal/modules/paypal',
```
wondering if this module should be last. Assuming modules are initialized in order and the paypal sdk has already loaded, the `paypal` module will init, try to call all the `handlers` (there will be none), and then move on. Then, the other modules will init and add themselves to the handlers list, which will never get called until/if `WORKAREA.initModules` is called again for something else. Right?

bencrouse replied:
That's a good point, but also the other modules need to be able to call the `ready` method to add their handlers. Maybe running the handlers should be in a `setTimeout(..., 0)` or something?

tubbo replied:
I should probably just modify `ready()` to call the function immediately if `paypalLoaded`.

mttdffy replied:
calling `ready` isn't a problem if they are loaded before the `WORKAREA.paypal` because they load, then initialize. So `WORKAREA.paypal.ready` will exist before its initialization method is called

tubbo replied:
I talked with Duffy on Slack about this a bit and we came to the conclusion that yes, this will most likely be a problem in the real world. To resolve this issue, the `init()` functions in the PayPal modules have been changed to just be calls to `WORKAREA.paypal.ready()`, which will call them on its own `init()` when PayPal has been loaded. This should also happen every time `initModules` is called, such as during an ajax request.

bencrouse replied:
Sounds good. Although it seems like to follow conventions, registering the PayPal stuff with `WORKAREA.paypal.ready` would just be part of the init for those modules:
```
  return {
    init: _.once(function () {
      WORKAREA.paypal.ready(function() {
        / * ... */
      });
    })
  }
```

tubbo replied:
That would still pose the problem of the external `<script>` not being loaded in time, as the module `init()` functions will be run _after_ PayPal's `init()`.

bencrouse replied:
Oh I see the confusion - if I call `WORKAREA.paypal.ready` _after_ PayPal has been loaded, it won't run what I've passed. I think that's a fair expectation of how a `ready` works. jQuery and the affirm API are two examples.

tubbo commented on app/assets/javascripts/workarea/storefront/paypal/modules/paypal.js:
```diff
@@ -0,0 +1,57 @@
+            if (!paypalLoaded && paypalExpectedToLoad && attempts <= 20) {
+                window.setTimeout(init, 2000);
+            } else if (paypalLoaded) {
+                handlers.forEach(function(handler) { handler($scope); });
```
we're now passing the `$scope` thru here so other paypal modules can use `WORKAREA.paypal.ready()` as a replacement for their own `init()` functions.

tubbo commented on app/assets/javascripts/workarea/storefront/paypal/modules/paypal_buttons.js:
```diff
@@ -49,23 +49,15 @@ WORKAREA.registerModule('paypalButtons', (function () {
-    return {
-        init: init
-    };
+    return {};
```
do I need this at the bottom, or can these modules just not return anything and that's OK?

bencrouse replied:
I dunno, I would probably do this as a just-in-case:
```
  return {
    init: $.noop
  }
```

tubbo replied:
Just confirmed that removing this doesn't seem to have an effect, so I'm doing it.

mttdffy commented on app/assets/javascripts/workarea/storefront/paypal/modules/paypal.js:
```diff
@@ -0,0 +1,62 @@
+            if (!paypalLoaded && paypalExpectedToLoad && attempts <= 20) {
+                window.setTimeout(init, 2000);
+            } else if (paypalLoaded) {
+                _.each(handlers, function(handler) { handler(); });
```
think handlers should be cleared out after this. Otherwise any secondary initialization of modules will refire those, which we wouldn't want, right?

tubbo replied:
If `initModules` gets fired twice, it would be on a different `$scope`, so yes we *would* want the handlers to fire again for a different DOM tree...

mttdffy replied:
nope. You are defining the handlers with a fixed scope. So you wouldn't want them to fire again, as they would always fire with the same scope as defined. Then the modules (button and hosted fields) would init again, adding a second handler for the same module with a different scope, or more likely just fire off immediately since PayPal would likely be loaded by the time another `initModules` was called.

Commit:
Use v3.5-stable branch to get Chartkick update

Need this to pass `bundler-audit`.

No changelog.

PAYPAL-7

tubbo commented:
`bundler-audit` will probably fail on this until workarea-commerce/workarea#483 is merged.

bencrouse commented:
@mttdffy you see the latest version?
